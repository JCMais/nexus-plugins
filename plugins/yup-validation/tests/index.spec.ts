import { execute, parse } from 'graphql'
import { makeSchema, mutationField, objectType, stringArg } from 'nexus'
import * as yup from 'yup'

import {
  yupValidationPlugin,
  MutationUserErrorNexusObjectType,
  YupValidationPluginConfig,
} from '../src'

const AddUserPayload = objectType({
  name: 'AddUserPayload',
  definition: (t) => {
    t.field('error', {
      type: MutationUserErrorNexusObjectType,
    })
  },
})

const AddUserMutation = mutationField((t) => {
  t.field('addUser', {
    type: AddUserPayload,
    args: {
      email: stringArg(),
    },
    // @ts-expect-error
    yup: {
      schema: yup.object({
        email: yup.string().email().min(10).required(),
      }),
    },
    // ...
  })
})

const CallAddUserMutation = parse(
  `
  mutation AddUser($email: String) {
    addUser(email: $email) {
      error {
        message
        details {
          path
          messages
        }
      }
    }
  }`,
)

const testSchema = (pluginConfig: YupValidationPluginConfig, outputs = false) =>
  makeSchema({
    outputs,
    types: [AddUserMutation],
    plugins: [yupValidationPlugin(pluginConfig)],
    nonNullDefaults: {
      input: false,
      output: false,
    },
  })

// TODO(jonathan): increase tests code coverage

describe('yupValidation', () => {
  describe('basic behavior', () => {
    it('should work correctly with default values', async () => {
      const schema = testSchema({})
      const data = await execute({
        schema,
        document: CallAddUserMutation,
        variableValues: {
          email: '',
        },
      })

      expect(data).toMatchSnapshot()
    })

    it('should work with abortEarly = true', async () => {
      const schema = testSchema({
        yup: {
          abortEarly: true,
        },
      })
      const data = await execute({
        schema,
        document: CallAddUserMutation,
        variableValues: {
          email: '',
        },
      })

      expect(data).toMatchSnapshot()
    })
  })
})
