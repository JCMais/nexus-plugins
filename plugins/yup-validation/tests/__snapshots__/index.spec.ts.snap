// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yupValidation basic behavior should work correctly with default values 1`] = `
Object {
  "data": Object {
    "addUser": Object {
      "error": Object {
        "details": Array [
          Object {
            "messages": Array [
              "email must be at least 10 characters",
              "email is a required field",
            ],
            "path": Array [
              "email",
            ],
          },
        ],
        "message": "2 errors occurred",
      },
    },
  },
}
`;

exports[`yupValidation basic behavior should work with abortEarly = true 1`] = `
Object {
  "data": Object {
    "addUser": Object {
      "error": Object {
        "details": Array [],
        "message": "email must be at least 10 characters",
      },
    },
  },
}
`;
